docs/proposals/0554-standadize-deployment-and-promotion.md

```diff
@@ -0,0 +1,813 @@
+   - Reduces duplication of shared definitions in Release objects
+   - Component-specific config can be embedded for faster access
+   - ComponentEnvSnapshot will be generated regardless
+
+## Implementation Phases
+
+### Phase 1: Core CRDs and Controllers
+- Implement Release and ReleasePin CRDs with both gitRef and content field support
+- Add validation webhooks to ensure exactly one of gitRef or content is specified
+- Implement ReleasePin controller with reconciliation logic for both approaches:
+  - Generate ComponentEnvSnapshot during reconciliation (not auto-generated)
+  - Support both git references and embedded content resolution
+  - Implement garbage collection for old ComponentEnvSnapshots
+- Support automatic Release creation on Component updates (first environment only)
+- Implement git fetching logic with caching (keyed by commit SHA)
+
+### Phase 2: UI/CLI Deployment and Promotion
+- Add deployment command to CLI: `choreo deploy --environment <environment>`
+- Add promotion command to CLI: `choreo promote --from <environment> --to <environment>`
+- Add "Deploy" button in OpenChoreo console UI with environment selection
+- Add "Promote" button in OpenChoreo console UI with source/target environment selection
+- Implement deployment and promotion audit logs
+- Add Release creation with embedded content for UI/CLI workflows
+
+### Phase 3: GitOps Support
+- Document GitOps workflows for promotions
+- Add validation webhooks to prevent invalid ReleasePin updates
+- Provide example GitOps repository structures
+- Document git authentication setup (SSH keys, personal access tokens)
+
+### Phase 4: Garbage Collection
+- Implement retention policies for UI/CLI mode
+- Add `choreo releases gc` command for GitOps mode
+- Add metrics for Release storage usage
+- Implement etcd size monitoring and alerting
+
+
+## Future Enhancements
+
+### 1. Content Compression for Embedded Releases
+
+Add transparent compression to reduce etcd storage footprint for embedded content:
+
+**Implementation approach:**
+- Add a `compressed: true` annotation to Release objects
+- Controller compresses embedded content using gzip before storing in etcd
+- Controller automatically decompresses during reconciliation
+- Expected compression ratios: 60-80% reduction (YAML compresses well)
+
+**Benefits:**
+- Reduces embedded Release size from ~50-200 KB to ~10-40 KB
+- Makes embedded content more competitive with git references in terms of storage
+- Particularly beneficial for UI/CLI workflows
+- Reduces etcd backup sizes and network transfer during replication
+
+**Example:**
+```yaml
+apiVersion: openchoreo.dev/v1alpha1
+kind: Release
+metadata:
+  name: checkout-service-v1.2.3
+  annotations:
+    openchoreo.dev/compression: gzip
+spec:
+  component:
+    name: checkout-service
+    content:
+      # Stored compressed in etcd, automatically decompressed on read
+      compressed: "H4sIAAAAAAAA/6tWKkktLlGyUlAqS8wpTtVRKi..."  # Base64-encoded gzip
```

Comment:
consider a faster/better alternative to b64 enc gzip (this is a double operation and there are better modern  things than gzip these days).

See some suggestions here: https://github.com/etcd-io/etcd/issues/1548#issuecomment-514497593